compiler/rustc_codegen_ssa/src/back/command.rs

```diff
@@ -137,44 +137,60 @@ impl Command {
-        // linker with smaller arguments.
-        //
-        // [1]: https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa
-        // [2]: https://devblogs.microsoft.com/oldnewthing/?p=41553
-
-        let estimated_command_line_len = self.args.iter().map(|a| a.len()).sum::<usize>();
-        estimated_command_line_len > 1024 * 6
+        if cfg!(windows) {
+            // Ok so on Windows to spawn a process is 32,768 characters in its
+            // command line [1]. Unfortunately we don't actually have access to that
+            // as it's calculated just before spawning. Instead we perform a
+            // poor-man's guess as to how long our command line will be. We're
+            // assuming here that we don't have to escape every character...
+            //
+            // Turns out though that `cmd.exe` has even smaller limits, 8192
+            // characters [2]. Linkers can often be batch scripts (for example
+            // Emscripten, Gecko's current build system) which means that we're
+            // running through batch scripts. These linkers often just forward
+            // arguments elsewhere (and maybe tack on more), so if we blow 8192
+            // bytes we'll typically cause them to blow as well.
+            //
+            // Basically as a result just perform an inflated estimate of what our
+            // command line will look like and test if it's > 8192 (we actually
+            // test against 6k to artificially inflate our estimate). If all else
+            // fails we'll fall back to the normal unix logic of testing the OS
+            // error code if we fail to spawn and automatically re-spawning the
+            // linker with smaller arguments.
+            //
+            // [1]: https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa
+            // [2]: https://devblogs.microsoft.com/oldnewthing/?p=41553
+            let estimated_command_line_len = self.args.iter().map(|a| a.len()).sum::<usize>();
+            estimated_command_line_len > 1024 * 6
+        } else if cfg!(unix) {
+            // On Unix the limits can be gargantuan anyway so we're pretty
+            // unlikely to hit them, but might still exceed it.
+            // We consult ARG_MAX here to get an estimate.
+            let ptr_size = mem::size_of::<usize>();
+            // arg + \0 + pointer
+            let args_size =
+                self.args.iter().map(|a| a.as_encoded_bytes().len() + 1 + ptr_size).sum::<usize>();
+            // key + `=` + value + \0 + pointer
+            let envs_size = self
+                .env
+                .iter()
+                .map(|(k, v)| k.as_encoded_bytes().len() + 1 + v.as_encoded_bytes().len() + 1 + ptr_size)
+                .sum::<usize>();
+            let arg_max = match unsafe { libc::sysconf(libc::_SC_ARG_MAX) } {
```

Comment:
This doesn't check overflow. Should we?

Reply:
Probably use `saturing_add`. Alternatively, checked addition that just falls back to go to the OS anyway if it would overflow.